// maplibre-gl-inspect/dist/maplibre-gl-inspect.cjs
'use strict';
/*!
 * maplibre-gl-inspect v0.0.6
 * Mapbox layer for average/interpolation heatmaps
 * (c) 2023 Vinayak Kulkarni<inbox.vinayak@gmail.com>
 * Released under the MIT License
 */ Object.defineProperties(exports, {
  __esModule: { value: !0 },
  [Symbol.toStringTag]: { value: 'Module' },
});
const ht = require('maplibre-gl'),
  An = require('randomcolor');
var nt =
  typeof globalThis < 'u'
    ? globalThis
    : typeof window < 'u'
    ? window
    : typeof global < 'u'
    ? global
    : typeof self < 'u'
    ? self
    : {};
function Pn(s) {
  return s && s.__esModule && Object.prototype.hasOwnProperty.call(s, 'default')
    ? s.default
    : s;
}
var F = {},
  In = {
    get exports() {
      return F;
    },
    set exports(s) {
      F = s;
    },
  },
  Jt;
function jn() {
  return (
    Jt ||
      ((Jt = 1),
      (function (s, r) {
        var o = 200,
          a = '__lodash_hash_undefined__',
          u = 1,
          C = 2,
          d = 9007199254740991,
          g = '[object Arguments]',
          y = '[object Array]',
          M = '[object AsyncFunction]',
          gt = '[object Boolean]',
          dt = '[object Date]',
          _t = '[object Error]',
          yt = '[object Function]',
          Qt = '[object GeneratorFunction]',
          U = '[object Map]',
          bt = '[object Number]',
          Vt = '[object Null]',
          D = '[object Object]',
          mt = '[object Promise]',
          te = '[object Proxy]',
          vt = '[object RegExp]',
          K = '[object Set]',
          wt = '[object String]',
          ee = '[object Symbol]',
          ne = '[object Undefined]',
          st = '[object WeakMap]',
          St = '[object ArrayBuffer]',
          J = '[object DataView]',
          re = '[object Float32Array]',
          se = '[object Float64Array]',
          ie = '[object Int8Array]',
          oe = '[object Int16Array]',
          ae = '[object Int32Array]',
          le = '[object Uint8Array]',
          pe = '[object Uint8ClampedArray]',
          ce = '[object Uint16Array]',
          ue = '[object Uint32Array]',
          fe = /[\\^$.*+?()[\]{}|]/g,
          he = /^\[object .+?Constructor\]$/,
          ge = /^(?:0|[1-9]\d*)$/,
          p = {};
        (p[re] =
          p[se] =
          p[ie] =
          p[oe] =
          p[ae] =
          p[le] =
          p[pe] =
          p[ce] =
          p[ue] =
            !0),
          (p[g] =
            p[y] =
            p[St] =
            p[gt] =
            p[J] =
            p[dt] =
            p[_t] =
            p[yt] =
            p[U] =
            p[bt] =
            p[D] =
            p[vt] =
            p[K] =
            p[wt] =
            p[st] =
              !1);
        var Ct = typeof nt == 'object' && nt && nt.Object === Object && nt,
          de =
            typeof self == 'object' && self && self.Object === Object && self,
          A = Ct || de || Function('return this')(),
          Tt = r && !r.nodeType && r,
          Ot = Tt && !0 && s && !s.nodeType && s,
          Mt = Ot && Ot.exports === Tt,
          it = Mt && Ct.process,
          At = (function () {
            try {
              return it && it.binding && it.binding('util');
            } catch {}
          })(),
          Pt = At && At.isTypedArray;
        function _e(t, e) {
          for (
            var n = -1, i = t == null ? 0 : t.length, c = 0, l = [];
            ++n < i;

          ) {
            var h = t[n];
            e(h, n, t) && (l[c++] = h);
          }
          return l;
        }
        function ye(t, e) {
          for (var n = -1, i = e.length, c = t.length; ++n < i; )
            t[c + n] = e[n];
          return t;
        }
        function be(t, e) {
          for (var n = -1, i = t == null ? 0 : t.length; ++n < i; )
            if (e(t[n], n, t)) return !0;
          return !1;
        }
        function me(t, e) {
          for (var n = -1, i = Array(t); ++n < t; ) i[n] = e(n);
          return i;
        }
        function ve(t) {
          return function (e) {
            return t(e);
          };
        }
        function we(t, e) {
          return t.has(e);
        }
        function Se(t, e) {
          return t?.[e];
        }
        function Ce(t) {
          var e = -1,
            n = Array(t.size);
          return (
            t.forEach(function (i, c) {
              n[++e] = [c, i];
            }),
            n
          );
        }
        function Te(t, e) {
          return function (n) {
            return t(e(n));
          };
        }
        function Oe(t) {
          var e = -1,
            n = Array(t.size);
          return (
            t.forEach(function (i) {
              n[++e] = i;
            }),
            n
          );
        }
        var Me = Array.prototype,
          Ae = Function.prototype,
          W = Object.prototype,
          ot = A['__core-js_shared__'],
          It = Ae.toString,
          O = W.hasOwnProperty,
          jt = (function () {
            var t = /[^.]+$/.exec((ot && ot.keys && ot.keys.IE_PROTO) || '');
            return t ? 'Symbol(src)_1.' + t : '';
          })(),
          Lt = W.toString,
          Pe = RegExp(
            '^' +
              It.call(O)
                .replace(fe, '\\$&')
                .replace(
                  /hasOwnProperty|(function).*?(?=\\\()| for .+?(?=\\\])/g,
                  '$1.*?',
                ) +
              '$',
          ),
          Et = Mt ? A.Buffer : void 0,
          X = A.Symbol,
          xt = A.Uint8Array,
          $t = W.propertyIsEnumerable,
          Ie = Me.splice,
          x = X ? X.toStringTag : void 0,
          Bt = Object.getOwnPropertySymbols,
          je = Et ? Et.isBuffer : void 0,
          Le = Te(Object.keys, Object),
          at = G(A, 'DataView'),
          q = G(A, 'Map'),
          lt = G(A, 'Promise'),
          pt = G(A, 'Set'),
          ct = G(A, 'WeakMap'),
          z = G(Object, 'create'),
          Ee = H(at),
          xe = H(q),
          $e = H(lt),
          Be = H(pt),
          He = H(ct),
          Ht = X ? X.prototype : void 0,
          ut = Ht ? Ht.valueOf : void 0;
        function $(t) {
          var e = -1,
            n = t == null ? 0 : t.length;
          for (this.clear(); ++e < n; ) {
            var i = t[e];
            this.set(i[0], i[1]);
          }
        }
        function Re() {
          (this.__data__ = z ? z(null) : {}), (this.size = 0);
        }
        function De(t) {
          var e = this.has(t) && delete this.__data__[t];
          return (this.size -= e ? 1 : 0), e;
        }
        function Ge(t) {
          var e = this.__data__;
          if (z) {
            var n = e[t];
            return n === a ? void 0 : n;
          }
          return O.call(e, t) ? e[t] : void 0;
        }
        function qe(t) {
          var e = this.__data__;
          return z ? e[t] !== void 0 : O.call(e, t);
        }
        function ze(t, e) {
          var n = this.__data__;
          return (
            (this.size += this.has(t) ? 0 : 1),
            (n[t] = z && e === void 0 ? a : e),
            this
          );
        }
        ($.prototype.clear = Re),
          ($.prototype.delete = De),
          ($.prototype.get = Ge),
          ($.prototype.has = qe),
          ($.prototype.set = ze);
        function P(t) {
          var e = -1,
            n = t == null ? 0 : t.length;
          for (this.clear(); ++e < n; ) {
            var i = t[e];
            this.set(i[0], i[1]);
          }
        }
        function Ne() {
          (this.__data__ = []), (this.size = 0);
        }
        function ke(t) {
          var e = this.__data__,
            n = Z(e, t);
          if (n < 0) return !1;
          var i = e.length - 1;
          return n == i ? e.pop() : Ie.call(e, n, 1), --this.size, !0;
        }
        function Fe(t) {
          var e = this.__data__,
            n = Z(e, t);
          return n < 0 ? void 0 : e[n][1];
        }
        function Ue(t) {
          return Z(this.__data__, t) > -1;
        }
        function Ke(t, e) {
          var n = this.__data__,
            i = Z(n, t);
          return i < 0 ? (++this.size, n.push([t, e])) : (n[i][1] = e), this;
        }
        (P.prototype.clear = Ne),
          (P.prototype.delete = ke),
          (P.prototype.get = Fe),
          (P.prototype.has = Ue),
          (P.prototype.set = Ke);
        function B(t) {
          var e = -1,
            n = t == null ? 0 : t.length;
          for (this.clear(); ++e < n; ) {
            var i = t[e];
            this.set(i[0], i[1]);
          }
        }
        function Je() {
          (this.size = 0),
            (this.__data__ = {
              hash: new $(),
              map: new (q || P)(),
              string: new $(),
            });
        }
        function We(t) {
          var e = Q(this, t).delete(t);
          return (this.size -= e ? 1 : 0), e;
        }
        function Xe(t) {
          return Q(this, t).get(t);
        }
        function Ye(t) {
          return Q(this, t).has(t);
        }
        function Ze(t, e) {
          var n = Q(this, t),
            i = n.size;
          return n.set(t, e), (this.size += n.size == i ? 0 : 1), this;
        }
        (B.prototype.clear = Je),
          (B.prototype.delete = We),
          (B.prototype.get = Xe),
          (B.prototype.has = Ye),
          (B.prototype.set = Ze);
        function Y(t) {
          var e = -1,
            n = t == null ? 0 : t.length;
          for (this.__data__ = new B(); ++e < n; ) this.add(t[e]);
        }
        function Qe(t) {
          return this.__data__.set(t, a), this;
        }
        function Ve(t) {
          return this.__data__.has(t);
        }
        (Y.prototype.add = Y.prototype.push = Qe), (Y.prototype.has = Ve);
        function j(t) {
          var e = (this.__data__ = new P(t));
          this.size = e.size;
        }
        function tn() {
          (this.__data__ = new P()), (this.size = 0);
        }
        function en(t) {
          var e = this.__data__,
            n = e.delete(t);
          return (this.size = e.size), n;
        }
        function nn(t) {
          return this.__data__.get(t);
        }
        function rn(t) {
          return this.__data__.has(t);
        }
        function sn(t, e) {
          var n = this.__data__;
          if (n instanceof P) {
            var i = n.__data__;
            if (!q || i.length < o - 1)
              return i.push([t, e]), (this.size = ++n.size), this;
            n = this.__data__ = new B(i);
          }
          return n.set(t, e), (this.size = n.size), this;
        }
        (j.prototype.clear = tn),
          (j.prototype.delete = en),
          (j.prototype.get = nn),
          (j.prototype.has = rn),
          (j.prototype.set = sn);
        function on(t, e) {
          var n = V(t),
            i = !n && wn(t),
            c = !n && !i && ft(t),
            l = !n && !i && !c && Ut(t),
            h = n || i || c || l,
            _ = h ? me(t.length, String) : [],
            b = _.length;
          for (var f in t)
            (e || O.call(t, f)) &&
              !(
                h &&
                (f == 'length' ||
                  (c && (f == 'offset' || f == 'parent')) ||
                  (l &&
                    (f == 'buffer' ||
                      f == 'byteLength' ||
                      f == 'byteOffset')) ||
                  _n(f, b))
              ) &&
              _.push(f);
          return _;
        }
        function Z(t, e) {
          for (var n = t.length; n--; ) if (zt(t[n][0], e)) return n;
          return -1;
        }
        function an(t, e, n) {
          var i = e(t);
          return V(t) ? i : ye(i, n(t));
        }
        function N(t) {
          return t == null
            ? t === void 0
              ? ne
              : Vt
            : x && x in Object(t)
            ? gn(t)
            : vn(t);
        }
        function Rt(t) {
          return k(t) && N(t) == g;
        }
        function Dt(t, e, n, i, c) {
          return t === e
            ? !0
            : t == null || e == null || (!k(t) && !k(e))
            ? t !== t && e !== e
            : ln(t, e, n, i, Dt, c);
        }
        function ln(t, e, n, i, c, l) {
          var h = V(t),
            _ = V(e),
            b = h ? y : L(t),
            f = _ ? y : L(e);
          (b = b == g ? D : b), (f = f == g ? D : f);
          var v = b == D,
            T = f == D,
            m = b == f;
          if (m && ft(t)) {
            if (!ft(e)) return !1;
            (h = !0), (v = !1);
          }
          if (m && !v)
            return (
              l || (l = new j()),
              h || Ut(t) ? Gt(t, e, n, i, c, l) : fn(t, e, b, n, i, c, l)
            );
          if (!(n & u)) {
            var w = v && O.call(t, '__wrapped__'),
              S = T && O.call(e, '__wrapped__');
            if (w || S) {
              var E = w ? t.value() : t,
                I = S ? e.value() : e;
              return l || (l = new j()), c(E, I, n, i, l);
            }
          }
          return m ? (l || (l = new j()), hn(t, e, n, i, c, l)) : !1;
        }
        function pn(t) {
          if (!Ft(t) || bn(t)) return !1;
          var e = Nt(t) ? Pe : he;
          return e.test(H(t));
        }
        function cn(t) {
          return k(t) && kt(t.length) && !!p[N(t)];
        }
        function un(t) {
          if (!mn(t)) return Le(t);
          var e = [];
          for (var n in Object(t))
            O.call(t, n) && n != 'constructor' && e.push(n);
          return e;
        }
        function Gt(t, e, n, i, c, l) {
          var h = n & u,
            _ = t.length,
            b = e.length;
          if (_ != b && !(h && b > _)) return !1;
          var f = l.get(t);
          if (f && l.get(e)) return f == e;
          var v = -1,
            T = !0,
            m = n & C ? new Y() : void 0;
          for (l.set(t, e), l.set(e, t); ++v < _; ) {
            var w = t[v],
              S = e[v];
            if (i) var E = h ? i(S, w, v, e, t, l) : i(w, S, v, t, e, l);
            if (E !== void 0) {
              if (E) continue;
              T = !1;
              break;
            }
            if (m) {
              if (
                !be(e, function (I, R) {
                  if (!we(m, R) && (w === I || c(w, I, n, i, l)))
                    return m.push(R);
                })
              ) {
                T = !1;
                break;
              }
            } else if (!(w === S || c(w, S, n, i, l))) {
              T = !1;
              break;
            }
          }
          return l.delete(t), l.delete(e), T;
        }
        function fn(t, e, n, i, c, l, h) {
          switch (n) {
            case J:
              if (t.byteLength != e.byteLength || t.byteOffset != e.byteOffset)
                return !1;
              (t = t.buffer), (e = e.buffer);
            case St:
              return !(
                t.byteLength != e.byteLength || !l(new xt(t), new xt(e))
              );
            case gt:
            case dt:
            case bt:
              return zt(+t, +e);
            case _t:
              return t.name == e.name && t.message == e.message;
            case vt:
            case wt:
              return t == e + '';
            case U:
              var _ = Ce;
            case K:
              var b = i & u;
              if ((_ || (_ = Oe), t.size != e.size && !b)) return !1;
              var f = h.get(t);
              if (f) return f == e;
              (i |= C), h.set(t, e);
              var v = Gt(_(t), _(e), i, c, l, h);
              return h.delete(t), v;
            case ee:
              if (ut) return ut.call(t) == ut.call(e);
          }
          return !1;
        }
        function hn(t, e, n, i, c, l) {
          var h = n & u,
            _ = qt(t),
            b = _.length,
            f = qt(e),
            v = f.length;
          if (b != v && !h) return !1;
          for (var T = b; T--; ) {
            var m = _[T];
            if (!(h ? m in e : O.call(e, m))) return !1;
          }
          var w = l.get(t);
          if (w && l.get(e)) return w == e;
          var S = !0;
          l.set(t, e), l.set(e, t);
          for (var E = h; ++T < b; ) {
            m = _[T];
            var I = t[m],
              R = e[m];
            if (i) var Kt = h ? i(R, I, m, e, t, l) : i(I, R, m, t, e, l);
            if (!(Kt === void 0 ? I === R || c(I, R, n, i, l) : Kt)) {
              S = !1;
              break;
            }
            E || (E = m == 'constructor');
          }
          if (S && !E) {
            var tt = t.constructor,
              et = e.constructor;
            tt != et &&
              'constructor' in t &&
              'constructor' in e &&
              !(
                typeof tt == 'function' &&
                tt instanceof tt &&
                typeof et == 'function' &&
                et instanceof et
              ) &&
              (S = !1);
          }
          return l.delete(t), l.delete(e), S;
        }
        function qt(t) {
          return an(t, Tn, dn);
        }
        function Q(t, e) {
          var n = t.__data__;
          return yn(e) ? n[typeof e == 'string' ? 'string' : 'hash'] : n.map;
        }
        function G(t, e) {
          var n = Se(t, e);
          return pn(n) ? n : void 0;
        }
        function gn(t) {
          var e = O.call(t, x),
            n = t[x];
          try {
            t[x] = void 0;
            var i = !0;
          } catch {}
          var c = Lt.call(t);
          return i && (e ? (t[x] = n) : delete t[x]), c;
        }
        var dn = Bt
            ? function (t) {
                return t == null
                  ? []
                  : ((t = Object(t)),
                    _e(Bt(t), function (e) {
                      return $t.call(t, e);
                    }));
              }
            : On,
          L = N;
        ((at && L(new at(new ArrayBuffer(1))) != J) ||
          (q && L(new q()) != U) ||
          (lt && L(lt.resolve()) != mt) ||
          (pt && L(new pt()) != K) ||
          (ct && L(new ct()) != st)) &&
          (L = function (t) {
            var e = N(t),
              n = e == D ? t.constructor : void 0,
              i = n ? H(n) : '';
            if (i)
              switch (i) {
                case Ee:
                  return J;
                case xe:
                  return U;
                case $e:
                  return mt;
                case Be:
                  return K;
                case He:
                  return st;
              }
            return e;
          });
        function _n(t, e) {
          return (
            (e = e ?? d),
            !!e &&
              (typeof t == 'number' || ge.test(t)) &&
              t > -1 &&
              t % 1 == 0 &&
              t < e
          );
        }
        function yn(t) {
          var e = typeof t;
          return e == 'string' ||
            e == 'number' ||
            e == 'symbol' ||
            e == 'boolean'
            ? t !== '__proto__'
            : t === null;
        }
        function bn(t) {
          return !!jt && jt in t;
        }
        function mn(t) {
          var e = t && t.constructor,
            n = (typeof e == 'function' && e.prototype) || W;
          return t === n;
        }
        function vn(t) {
          return Lt.call(t);
        }
        function H(t) {
          if (t != null) {
            try {
              return It.call(t);
            } catch {}
            try {
              return t + '';
            } catch {}
          }
          return '';
        }
        function zt(t, e) {
          return t === e || (t !== t && e !== e);
        }
        var wn = Rt(
            (function () {
              return arguments;
            })(),
          )
            ? Rt
            : function (t) {
                return k(t) && O.call(t, 'callee') && !$t.call(t, 'callee');
              },
          V = Array.isArray;
        function Sn(t) {
          return t != null && kt(t.length) && !Nt(t);
        }
        var ft = je || Mn;
        function Cn(t, e) {
          return Dt(t, e);
        }
        function Nt(t) {
          if (!Ft(t)) return !1;
          var e = N(t);
          return e == yt || e == Qt || e == M || e == te;
        }
        function kt(t) {
          return typeof t == 'number' && t > -1 && t % 1 == 0 && t <= d;
        }
        function Ft(t) {
          var e = typeof t;
          return t != null && (e == 'object' || e == 'function');
        }
        function k(t) {
          return t != null && typeof t == 'object';
        }
        var Ut = Pt ? ve(Pt) : cn;
        function Tn(t) {
          return Sn(t) ? on(t) : un(t);
        }
        function On() {
          return [];
        }
        function Mn() {
          return !1;
        }
        s.exports = Cn;
      })(In, F)),
    F
  );
}
var Ln = jn();
const En = Pn(Ln),
  Wt = (s, r, o) => {
    const a = {
      id: [r, o, 'circle'].join('_'),
      source: r,
      type: 'circle',
      paint: { 'circle-color': s, 'circle-radius': 2 },
      filter: ['==', '$type', 'Point'],
    };
    return o && (a['source-layer'] = o), a;
  },
  Xt = (s, r, o, a) => {
    const u = {
      id: [o, a, 'polygon'].join('_'),
      source: o,
      type: 'fill',
      paint: { 'fill-antialias': !0, 'fill-color': s, 'fill-outline-color': r },
      filter: ['==', '$type', 'Polygon'],
    };
    return a && (u['source-layer'] = a), u;
  },
  Yt = (s, r, o) => {
    const a = {
      id: [r, o, 'line'].join('_'),
      source: r,
      layout: { 'line-join': 'round', 'line-cap': 'round' },
      type: 'line',
      paint: { 'line-color': s },
      filter: ['==', '$type', 'LineString'],
    };
    return o && (a['source-layer'] = o), a;
  },
  xn = (s, r) => {
    const o = [],
      a = [],
      u = [],
      C = (d) => {
        const g = r.bind(null, d);
        return {
          circle: g(0.8),
          line: g(0.6),
          polygon: g(0.3),
          polygonOutline: g(0.6),
          default: g(1),
        };
      };
    return (
      Object.keys(s).forEach((d) => {
        const g = s[`${d}`];
        if (!g || g.length === 0) {
          const y = C(d);
          a.push(Wt(y.circle, d)),
            u.push(Yt(y.line, d)),
            o.push(Xt(y.polygon, y.polygonOutline, d));
        } else
          g.forEach((y) => {
            const M = C(y);
            a.push(Wt(M.circle, d, y)),
              u.push(Yt(M.line, d, y)),
              o.push(Xt(M.polygon, M.polygonOutline, d, y));
          });
      }),
      [...o, ...u, ...a]
    );
  },
  $n = (s, r, o) => {
    o = Object.assign({ backgroundColor: '#fff' }, o);
    const a = {
        id: 'background',
        type: 'background',
        paint: { 'background-color': o.backgroundColor },
      },
      u = {};
    return (
      Object.keys(s.sources).forEach((C) => {
        const d = s.sources[`${C}`];
        (d.type === 'vector' || d.type === 'geojson') && (u[`${C}`] = d);
      }),
      { ...s, layers: [a, ...r], sources: u }
    );
  };
class Bn {
  _btn;
  elem;
  constructor(r) {
    (r = Object.assign({ show: !0, onToggle: () => {} }, r)),
      (this._btn = this.button()),
      (this._btn.onclick = r.onToggle),
      (this.elem = this.container(this._btn, r.show));
  }
  container(r, o) {
    const a = document.createElement('div');
    return (
      (a.className = 'maplibregl-ctrl maplibregl-ctrl-group'),
      a.appendChild(r),
      o || (a.style.display = 'none'),
      a
    );
  }
  button() {
    const r = document.createElement('button');
    return (
      (r.className = 'maplibregl-ctrl-icon maplibregl-ctrl-inspect'),
      (r.type = 'button'),
      (r.ariaLabel = 'Inspect'),
      r
    );
  }
  setInspectIcon() {
    this._btn.className = 'maplibregl-ctrl-icon maplibregl-ctrl-inspect';
  }
  setMapIcon() {
    this._btn.className = 'maplibregl-ctrl-icon maplibregl-ctrl-map';
  }
}
const Hn = (s) =>
    typeof s > 'u' || s === null
      ? s
      : s instanceof Date
      ? s.toLocaleString()
      : typeof s == 'object' || typeof s == 'number' || typeof s == 'string'
      ? s.toString()
      : s,
  Zt = (s, r) =>
    `<div class="maplibregl-inspect-property"><div class="maplibregl-inspect-property-name">${s}</div><div class="maplibregl-inspect-property-value">${Hn(
      r,
    )}</div></div>`,
  Rn = (s) => `<div class="maplibregl-inspect-layer">${s}</div>`,
  Dn = (s) => {
    const r = Rn(s.layer['source-layer'] || s.layer.source),
      o = Zt('$type', s.geometry.type),
      a = Object.keys(s.properties).map((u) => Zt(u, s.properties[`${u}`]));
    return [r, o].concat(a).join('');
  },
  Gn = (s) =>
    s
      .map((r) => `<div class="maplibregl-inspect-feature">${Dn(r)}</div>`)
      .join(''),
  qn = (s) => `<div class="maplibregl-inspect-popup">${Gn(s)}</div>`,
  zn = (s, r) => {
    let o = 'bright',
      a;
    return (
      /water|ocean|lake|sea|river/.test(s) && (a = 'blue'),
      /state|country|place/.test(s) && (a = 'pink'),
      /road|highway|transport|streets/.test(s) && (a = 'orange'),
      /contour|building/.test(s) && (a = 'monochrome'),
      /building/.test(s) && (o = 'dark'),
      /contour|landuse/.test(s) && (a = 'yellow'),
      /wood|forest|park|landcover|land/.test(s) && (a = 'green'),
      `rgba(${`${An({ luminosity: o, hue: a, seed: s, format: 'rgbArray' })},${
        r || 1
      }`})`
    );
  },
  Nn = (s) =>
    !!(s.metadata && 'maplibregl-inspect:inspect' in Object.keys(s.metadata)),
  kn = (s) => ({
    ...s,
    metadata: Object.assign({}, s.metadata, {
      'maplibregl-inspect:inspect': !0,
    }),
  });
class rt {
  _map;
  _popup;
  _popupBlocked = !1;
  _showInspectMap;
  _originalStyle;
  _toggle;
  options;
  sources;
  assignLayerColor;
  constructor(r) {
    if (!(this instanceof rt))
      throw new Error(
        'MaplibreInspect needs to be called with the new keyword',
      );
    let o = null;
    ht
      ? (o = new ht.Popup({ closeButton: !1, closeOnClick: !1 }))
      : r.popup ||
        console.error(
          'Maplibre GL JS can not be found. Make sure to include it or pass an initialized MaplibreGL Popup to MaplibreInspect if you are using moduleis.',
        ),
      (this.options = {
        showInspectMap: !1,
        showInspectButton: !0,
        showInspectMapPopup: !0,
        showMapPopup: !1,
        showMapPopupOnHover: !0,
        showInspectMapPopupOnHover: !0,
        blockHoverPopupOnClick: !1,
        backgroundColor: '#fff',
        assignLayerColor: zn,
        buildInspectStyle: $n,
        renderPopup: qn,
        popup: o,
        selectThreshold: 5,
        useInspectStyle: !0,
        queryParameters: {},
        sources: {},
        toggleCallback(a) {
          console.log('Inspector status?: ', a);
        },
        ...r,
      }),
      (this.sources = this.options.sources),
      (this.assignLayerColor = this.options.assignLayerColor),
      (this.toggleInspector = this.toggleInspector.bind(this)),
      (this._popup = this.options.popup),
      (this._popupBlocked = !1),
      (this._showInspectMap = this.options.showInspectMap),
      (this._onSourceChange = this._onSourceChange.bind(this)),
      (this._onMouseMove = this._onMouseMove.bind(this)),
      (this._onRightClick = this._onRightClick.bind(this)),
      (this._onStyleChange = this._onStyleChange.bind(this)),
      (this._originalStyle = null),
      (this._toggle = new Bn({
        show: this.options.showInspectButton,
        onToggle: this.toggleInspector.bind(this),
      }));
  }
  _inspectStyle() {
    let r = this._map?.getStyle();
    if (this._map) {
      const o = xn(this.sources, this.assignLayerColor);
      r = this.options.buildInspectStyle(this._map.getStyle(), o, {
        backgroundColor: this.options.backgroundColor,
      });
    }
    return r;
  }
  _onSourceChange(r) {
    const o = this.sources;
    if (this._map) {
      const a = this._map,
        u = a.getStyle(),
        C = Object.keys(u.sources),
        d = Object.assign({}, o);
      r.isSourceLoaded &&
        (Object.keys(a.style.sourceCaches).forEach((g) => {
          const y = a.style.sourceCaches[`${g}`] || { _source: {} },
            M = y._source.vectorLayerIds;
          M
            ? (o[`${g}`] = M)
            : y._source.type === 'geojson' && (o[`${g}`] = []);
        }),
        Object.keys(o).forEach((g) => {
          C.indexOf(g) === -1 && delete o[`${g}`];
        }),
        !En(d, o) && Object.keys(o).length > 0 && this.render());
    }
  }
  _onStyleChange() {
    const r = this._map?.getStyle();
    r && !Nn(r) && (this._originalStyle = r);
  }
  _onRightClick() {
    !this.options.showMapPopupOnHover &&
      !this.options.showInspectMapPopupOnHover &&
      !this.options.blockHoverPopupOnClick &&
      this._popup &&
      this._popup.remove();
  }
  _onMouseMove(r) {
    if (this._showInspectMap) {
      if (
        !this.options.showInspectMapPopup ||
        (r.type === 'mousemove' && !this.options.showInspectMapPopupOnHover)
      )
        return;
      r.type === 'click' &&
        this.options.showInspectMapPopupOnHover &&
        this.options.blockHoverPopupOnClick &&
        (this._popupBlocked = !this._popupBlocked);
    } else {
      if (
        !this.options.showMapPopup ||
        (r.type === 'mousemove' && !this.options.showMapPopupOnHover)
      )
        return;
      r.type === 'click' &&
        this.options.showMapPopupOnHover &&
        this.options.blockHoverPopupOnClick &&
        (this._popupBlocked = !this._popupBlocked);
    }
    if (!this._popupBlocked && this._map) {
      let o;
      this.options.selectThreshold === 0
        ? (o = r.point)
        : (o = [
            [
              r.point.x - this.options.selectThreshold,
              r.point.y + this.options.selectThreshold,
            ],
            [
              r.point.x + this.options.selectThreshold,
              r.point.y - this.options.selectThreshold,
            ],
          ]);
      const a =
        this._map.queryRenderedFeatures(o, this.options.queryParameters) || [];
      if (
        ((this._map.getCanvas().style.cursor = a.length ? 'pointer' : ''),
        a.length > 0 && this._popup instanceof ht.Popup)
      ) {
        this._popup.setLngLat(r.lngLat);
        const u = this.options.renderPopup(a);
        typeof u == 'string'
          ? this._popup.setHTML(u)
          : this._popup.setDOMContent(u),
          this._popup.addTo(this._map);
      } else this._popup?.remove();
    }
  }
  toggleInspector() {
    (this._showInspectMap = !this._showInspectMap),
      this.options.toggleCallback(this._showInspectMap),
      this.render();
  }
  render() {
    if (this._showInspectMap) {
      if (this.options.useInspectStyle) {
        const r = this._inspectStyle();
        this._map?.setStyle(kn(r));
      }
      this._toggle.setMapIcon();
    }
    !this._showInspectMap &&
      this._originalStyle &&
      (this.options.useInspectStyle && this._map?.setStyle(this._originalStyle),
      this._popup && this._popup.remove(),
      this._toggle.setInspectIcon());
  }
  onAdd(r) {
    return (
      (this._map = r),
      Object.keys(this.sources).length === 0 &&
        r.on('sourcedata', this._onSourceChange),
      r.on('styledata', this._onStyleChange),
      r.on('load', this._onStyleChange),
      r.on('mousemove', this._onMouseMove),
      r.on('click', this._onMouseMove),
      r.on('contextmenu', this._onRightClick),
      this._toggle.elem
    );
  }
  onRemove() {
    this._map?.off('styledata', this._onStyleChange),
      this._map?.off('load', this._onStyleChange),
      this._map?.off('sourcedata', this._onSourceChange),
      this._map?.off('mousemove', this._onMouseMove),
      this._map?.off('click', this._onMouseMove),
      this._map?.off('contextmenu', this._onRightClick);
    const r = this._toggle.elem;
    r.parentNode?.removeChild(r), (this._map = void 0);
  }
}
exports.MaplibreInspect = rt;
exports.default = rt;
