// maplibre-gl-inspect/dist/maplibre-gl-inspect.umd.cjs
(function (O, $) {
  typeof exports == 'object' && typeof module < 'u'
    ? $(exports, require('maplibre-gl'), require('randomcolor'))
    : typeof define == 'function' && define.amd
    ? define(['exports', 'maplibre-gl', 'randomcolor'], $)
    : ((O = typeof globalThis < 'u' ? globalThis : O || self),
      $(
        (O.MaplibreGlInspect = O.MaplibreGlInspect || {}),
        O.maplibregl,
        O.randomColor,
      ));
})(this, function (O, $, Ve) {
  'use strict';
  /*!
   * maplibre-gl-inspect v0.0.6
   * Mapbox layer for average/interpolation heatmaps
   * (c) 2023 Vinayak Kulkarni<inbox.vinayak@gmail.com>
   * Released under the MIT License
   */ const Un = '';
  var J =
    typeof globalThis < 'u'
      ? globalThis
      : typeof window < 'u'
      ? window
      : typeof global < 'u'
      ? global
      : typeof self < 'u'
      ? self
      : {};
  function et(s) {
    return s &&
      s.__esModule &&
      Object.prototype.hasOwnProperty.call(s, 'default')
      ? s.default
      : s;
  }
  var N = {},
    tt = {
      get exports() {
        return N;
      },
      set exports(s) {
        N = s;
      },
    },
    de;
  function nt() {
    return (
      de ||
        ((de = 1),
        (function (s, r) {
          var o = 200,
            a = '__lodash_hash_undefined__',
            u = 1,
            C = 2,
            d = 9007199254740991,
            g = '[object Arguments]',
            y = '[object Array]',
            A = '[object AsyncFunction]',
            ve = '[object Boolean]',
            we = '[object Date]',
            Se = '[object Error]',
            Ce = '[object Function]',
            _t = '[object GeneratorFunction]',
            X = '[object Map]',
            Te = '[object Number]',
            yt = '[object Null]',
            q = '[object Object]',
            Oe = '[object Promise]',
            bt = '[object Proxy]',
            Me = '[object RegExp]',
            Y = '[object Set]',
            Ae = '[object String]',
            mt = '[object Symbol]',
            vt = '[object Undefined]',
            oe = '[object WeakMap]',
            Ie = '[object ArrayBuffer]',
            Z = '[object DataView]',
            wt = '[object Float32Array]',
            St = '[object Float64Array]',
            Ct = '[object Int8Array]',
            Tt = '[object Int16Array]',
            Ot = '[object Int32Array]',
            Mt = '[object Uint8Array]',
            At = '[object Uint8ClampedArray]',
            It = '[object Uint16Array]',
            Pt = '[object Uint32Array]',
            jt = /[\\^$.*+?()[\]{}|]/g,
            Lt = /^\[object .+?Constructor\]$/,
            Et = /^(?:0|[1-9]\d*)$/,
            p = {};
          (p[wt] =
            p[St] =
            p[Ct] =
            p[Tt] =
            p[Ot] =
            p[Mt] =
            p[At] =
            p[It] =
            p[Pt] =
              !0),
            (p[g] =
              p[y] =
              p[Ie] =
              p[ve] =
              p[Z] =
              p[we] =
              p[Se] =
              p[Ce] =
              p[X] =
              p[Te] =
              p[q] =
              p[Me] =
              p[Y] =
              p[Ae] =
              p[oe] =
                !1);
          var Pe = typeof J == 'object' && J && J.Object === Object && J,
            xt =
              typeof self == 'object' && self && self.Object === Object && self,
            I = Pe || xt || Function('return this')(),
            je = r && !r.nodeType && r,
            Le = je && !0 && s && !s.nodeType && s,
            Ee = Le && Le.exports === je,
            ae = Ee && Pe.process,
            xe = (function () {
              try {
                return ae && ae.binding && ae.binding('util');
              } catch {}
            })(),
            $e = xe && xe.isTypedArray;
          function $t(e, t) {
            for (
              var n = -1, i = e == null ? 0 : e.length, c = 0, l = [];
              ++n < i;

            ) {
              var h = e[n];
              t(h, n, e) && (l[c++] = h);
            }
            return l;
          }
          function Bt(e, t) {
            for (var n = -1, i = t.length, c = e.length; ++n < i; )
              e[c + n] = t[n];
            return e;
          }
          function Ht(e, t) {
            for (var n = -1, i = e == null ? 0 : e.length; ++n < i; )
              if (t(e[n], n, e)) return !0;
            return !1;
          }
          function Rt(e, t) {
            for (var n = -1, i = Array(e); ++n < e; ) i[n] = t(n);
            return i;
          }
          function Gt(e) {
            return function (t) {
              return e(t);
            };
          }
          function Dt(e, t) {
            return e.has(t);
          }
          function qt(e, t) {
            return e?.[t];
          }
          function zt(e) {
            var t = -1,
              n = Array(e.size);
            return (
              e.forEach(function (i, c) {
                n[++t] = [c, i];
              }),
              n
            );
          }
          function Nt(e, t) {
            return function (n) {
              return e(t(n));
            };
          }
          function kt(e) {
            var t = -1,
              n = Array(e.size);
            return (
              e.forEach(function (i) {
                n[++t] = i;
              }),
              n
            );
          }
          var Ft = Array.prototype,
            Ut = Function.prototype,
            Q = Object.prototype,
            le = I['__core-js_shared__'],
            Be = Ut.toString,
            M = Q.hasOwnProperty,
            He = (function () {
              var e = /[^.]+$/.exec((le && le.keys && le.keys.IE_PROTO) || '');
              return e ? 'Symbol(src)_1.' + e : '';
            })(),
            Re = Q.toString,
            Kt = RegExp(
              '^' +
                Be.call(M)
                  .replace(jt, '\\$&')
                  .replace(
                    /hasOwnProperty|(function).*?(?=\\\()| for .+?(?=\\\])/g,
                    '$1.*?',
                  ) +
                '$',
            ),
            Ge = Ee ? I.Buffer : void 0,
            V = I.Symbol,
            De = I.Uint8Array,
            qe = Q.propertyIsEnumerable,
            Jt = Ft.splice,
            B = V ? V.toStringTag : void 0,
            ze = Object.getOwnPropertySymbols,
            Wt = Ge ? Ge.isBuffer : void 0,
            Xt = Nt(Object.keys, Object),
            pe = z(I, 'DataView'),
            k = z(I, 'Map'),
            ce = z(I, 'Promise'),
            ue = z(I, 'Set'),
            fe = z(I, 'WeakMap'),
            F = z(Object, 'create'),
            Yt = G(pe),
            Zt = G(k),
            Qt = G(ce),
            Vt = G(ue),
            en = G(fe),
            Ne = V ? V.prototype : void 0,
            he = Ne ? Ne.valueOf : void 0;
          function H(e) {
            var t = -1,
              n = e == null ? 0 : e.length;
            for (this.clear(); ++t < n; ) {
              var i = e[t];
              this.set(i[0], i[1]);
            }
          }
          function tn() {
            (this.__data__ = F ? F(null) : {}), (this.size = 0);
          }
          function nn(e) {
            var t = this.has(e) && delete this.__data__[e];
            return (this.size -= t ? 1 : 0), t;
          }
          function rn(e) {
            var t = this.__data__;
            if (F) {
              var n = t[e];
              return n === a ? void 0 : n;
            }
            return M.call(t, e) ? t[e] : void 0;
          }
          function sn(e) {
            var t = this.__data__;
            return F ? t[e] !== void 0 : M.call(t, e);
          }
          function on(e, t) {
            var n = this.__data__;
            return (
              (this.size += this.has(e) ? 0 : 1),
              (n[e] = F && t === void 0 ? a : t),
              this
            );
          }
          (H.prototype.clear = tn),
            (H.prototype.delete = nn),
            (H.prototype.get = rn),
            (H.prototype.has = sn),
            (H.prototype.set = on);
          function P(e) {
            var t = -1,
              n = e == null ? 0 : e.length;
            for (this.clear(); ++t < n; ) {
              var i = e[t];
              this.set(i[0], i[1]);
            }
          }
          function an() {
            (this.__data__ = []), (this.size = 0);
          }
          function ln(e) {
            var t = this.__data__,
              n = te(t, e);
            if (n < 0) return !1;
            var i = t.length - 1;
            return n == i ? t.pop() : Jt.call(t, n, 1), --this.size, !0;
          }
          function pn(e) {
            var t = this.__data__,
              n = te(t, e);
            return n < 0 ? void 0 : t[n][1];
          }
          function cn(e) {
            return te(this.__data__, e) > -1;
          }
          function un(e, t) {
            var n = this.__data__,
              i = te(n, e);
            return i < 0 ? (++this.size, n.push([e, t])) : (n[i][1] = t), this;
          }
          (P.prototype.clear = an),
            (P.prototype.delete = ln),
            (P.prototype.get = pn),
            (P.prototype.has = cn),
            (P.prototype.set = un);
          function R(e) {
            var t = -1,
              n = e == null ? 0 : e.length;
            for (this.clear(); ++t < n; ) {
              var i = e[t];
              this.set(i[0], i[1]);
            }
          }
          function fn() {
            (this.size = 0),
              (this.__data__ = {
                hash: new H(),
                map: new (k || P)(),
                string: new H(),
              });
          }
          function hn(e) {
            var t = ne(this, e).delete(e);
            return (this.size -= t ? 1 : 0), t;
          }
          function gn(e) {
            return ne(this, e).get(e);
          }
          function dn(e) {
            return ne(this, e).has(e);
          }
          function _n(e, t) {
            var n = ne(this, e),
              i = n.size;
            return n.set(e, t), (this.size += n.size == i ? 0 : 1), this;
          }
          (R.prototype.clear = fn),
            (R.prototype.delete = hn),
            (R.prototype.get = gn),
            (R.prototype.has = dn),
            (R.prototype.set = _n);
          function ee(e) {
            var t = -1,
              n = e == null ? 0 : e.length;
            for (this.__data__ = new R(); ++t < n; ) this.add(e[t]);
          }
          function yn(e) {
            return this.__data__.set(e, a), this;
          }
          function bn(e) {
            return this.__data__.has(e);
          }
          (ee.prototype.add = ee.prototype.push = yn), (ee.prototype.has = bn);
          function L(e) {
            var t = (this.__data__ = new P(e));
            this.size = t.size;
          }
          function mn() {
            (this.__data__ = new P()), (this.size = 0);
          }
          function vn(e) {
            var t = this.__data__,
              n = t.delete(e);
            return (this.size = t.size), n;
          }
          function wn(e) {
            return this.__data__.get(e);
          }
          function Sn(e) {
            return this.__data__.has(e);
          }
          function Cn(e, t) {
            var n = this.__data__;
            if (n instanceof P) {
              var i = n.__data__;
              if (!k || i.length < o - 1)
                return i.push([e, t]), (this.size = ++n.size), this;
              n = this.__data__ = new R(i);
            }
            return n.set(e, t), (this.size = n.size), this;
          }
          (L.prototype.clear = mn),
            (L.prototype.delete = vn),
            (L.prototype.get = wn),
            (L.prototype.has = Sn),
            (L.prototype.set = Cn);
          function Tn(e, t) {
            var n = re(e),
              i = !n && Dn(e),
              c = !n && !i && ge(e),
              l = !n && !i && !c && Ze(e),
              h = n || i || c || l,
              _ = h ? Rt(e.length, String) : [],
              b = _.length;
            for (var f in e)
              (t || M.call(e, f)) &&
                !(
                  h &&
                  (f == 'length' ||
                    (c && (f == 'offset' || f == 'parent')) ||
                    (l &&
                      (f == 'buffer' ||
                        f == 'byteLength' ||
                        f == 'byteOffset')) ||
                    $n(f, b))
                ) &&
                _.push(f);
            return _;
          }
          function te(e, t) {
            for (var n = e.length; n--; ) if (Je(e[n][0], t)) return n;
            return -1;
          }
          function On(e, t, n) {
            var i = t(e);
            return re(e) ? i : Bt(i, n(e));
          }
          function U(e) {
            return e == null
              ? e === void 0
                ? vt
                : yt
              : B && B in Object(e)
              ? En(e)
              : Gn(e);
          }
          function ke(e) {
            return K(e) && U(e) == g;
          }
          function Fe(e, t, n, i, c) {
            return e === t
              ? !0
              : e == null || t == null || (!K(e) && !K(t))
              ? e !== e && t !== t
              : Mn(e, t, n, i, Fe, c);
          }
          function Mn(e, t, n, i, c, l) {
            var h = re(e),
              _ = re(t),
              b = h ? y : E(e),
              f = _ ? y : E(t);
            (b = b == g ? q : b), (f = f == g ? q : f);
            var v = b == q,
              T = f == q,
              m = b == f;
            if (m && ge(e)) {
              if (!ge(t)) return !1;
              (h = !0), (v = !1);
            }
            if (m && !v)
              return (
                l || (l = new L()),
                h || Ze(e) ? Ue(e, t, n, i, c, l) : jn(e, t, b, n, i, c, l)
              );
            if (!(n & u)) {
              var w = v && M.call(e, '__wrapped__'),
                S = T && M.call(t, '__wrapped__');
              if (w || S) {
                var x = w ? e.value() : e,
                  j = S ? t.value() : t;
                return l || (l = new L()), c(x, j, n, i, l);
              }
            }
            return m ? (l || (l = new L()), Ln(e, t, n, i, c, l)) : !1;
          }
          function An(e) {
            if (!Ye(e) || Hn(e)) return !1;
            var t = We(e) ? Kt : Lt;
            return t.test(G(e));
          }
          function In(e) {
            return K(e) && Xe(e.length) && !!p[U(e)];
          }
          function Pn(e) {
            if (!Rn(e)) return Xt(e);
            var t = [];
            for (var n in Object(e))
              M.call(e, n) && n != 'constructor' && t.push(n);
            return t;
          }
          function Ue(e, t, n, i, c, l) {
            var h = n & u,
              _ = e.length,
              b = t.length;
            if (_ != b && !(h && b > _)) return !1;
            var f = l.get(e);
            if (f && l.get(t)) return f == t;
            var v = -1,
              T = !0,
              m = n & C ? new ee() : void 0;
            for (l.set(e, t), l.set(t, e); ++v < _; ) {
              var w = e[v],
                S = t[v];
              if (i) var x = h ? i(S, w, v, t, e, l) : i(w, S, v, e, t, l);
              if (x !== void 0) {
                if (x) continue;
                T = !1;
                break;
              }
              if (m) {
                if (
                  !Ht(t, function (j, D) {
                    if (!Dt(m, D) && (w === j || c(w, j, n, i, l)))
                      return m.push(D);
                  })
                ) {
                  T = !1;
                  break;
                }
              } else if (!(w === S || c(w, S, n, i, l))) {
                T = !1;
                break;
              }
            }
            return l.delete(e), l.delete(t), T;
          }
          function jn(e, t, n, i, c, l, h) {
            switch (n) {
              case Z:
                if (
                  e.byteLength != t.byteLength ||
                  e.byteOffset != t.byteOffset
                )
                  return !1;
                (e = e.buffer), (t = t.buffer);
              case Ie:
                return !(
                  e.byteLength != t.byteLength || !l(new De(e), new De(t))
                );
              case ve:
              case we:
              case Te:
                return Je(+e, +t);
              case Se:
                return e.name == t.name && e.message == t.message;
              case Me:
              case Ae:
                return e == t + '';
              case X:
                var _ = zt;
              case Y:
                var b = i & u;
                if ((_ || (_ = kt), e.size != t.size && !b)) return !1;
                var f = h.get(e);
                if (f) return f == t;
                (i |= C), h.set(e, t);
                var v = Ue(_(e), _(t), i, c, l, h);
                return h.delete(e), v;
              case mt:
                if (he) return he.call(e) == he.call(t);
            }
            return !1;
          }
          function Ln(e, t, n, i, c, l) {
            var h = n & u,
              _ = Ke(e),
              b = _.length,
              f = Ke(t),
              v = f.length;
            if (b != v && !h) return !1;
            for (var T = b; T--; ) {
              var m = _[T];
              if (!(h ? m in t : M.call(t, m))) return !1;
            }
            var w = l.get(e);
            if (w && l.get(t)) return w == t;
            var S = !0;
            l.set(e, t), l.set(t, e);
            for (var x = h; ++T < b; ) {
              m = _[T];
              var j = e[m],
                D = t[m];
              if (i) var Qe = h ? i(D, j, m, t, e, l) : i(j, D, m, e, t, l);
              if (!(Qe === void 0 ? j === D || c(j, D, n, i, l) : Qe)) {
                S = !1;
                break;
              }
              x || (x = m == 'constructor');
            }
            if (S && !x) {
              var se = e.constructor,
                ie = t.constructor;
              se != ie &&
                'constructor' in e &&
                'constructor' in t &&
                !(
                  typeof se == 'function' &&
                  se instanceof se &&
                  typeof ie == 'function' &&
                  ie instanceof ie
                ) &&
                (S = !1);
            }
            return l.delete(e), l.delete(t), S;
          }
          function Ke(e) {
            return On(e, Nn, xn);
          }
          function ne(e, t) {
            var n = e.__data__;
            return Bn(t) ? n[typeof t == 'string' ? 'string' : 'hash'] : n.map;
          }
          function z(e, t) {
            var n = qt(e, t);
            return An(n) ? n : void 0;
          }
          function En(e) {
            var t = M.call(e, B),
              n = e[B];
            try {
              e[B] = void 0;
              var i = !0;
            } catch {}
            var c = Re.call(e);
            return i && (t ? (e[B] = n) : delete e[B]), c;
          }
          var xn = ze
              ? function (e) {
                  return e == null
                    ? []
                    : ((e = Object(e)),
                      $t(ze(e), function (t) {
                        return qe.call(e, t);
                      }));
                }
              : kn,
            E = U;
          ((pe && E(new pe(new ArrayBuffer(1))) != Z) ||
            (k && E(new k()) != X) ||
            (ce && E(ce.resolve()) != Oe) ||
            (ue && E(new ue()) != Y) ||
            (fe && E(new fe()) != oe)) &&
            (E = function (e) {
              var t = U(e),
                n = t == q ? e.constructor : void 0,
                i = n ? G(n) : '';
              if (i)
                switch (i) {
                  case Yt:
                    return Z;
                  case Zt:
                    return X;
                  case Qt:
                    return Oe;
                  case Vt:
                    return Y;
                  case en:
                    return oe;
                }
              return t;
            });
          function $n(e, t) {
            return (
              (t = t ?? d),
              !!t &&
                (typeof e == 'number' || Et.test(e)) &&
                e > -1 &&
                e % 1 == 0 &&
                e < t
            );
          }
          function Bn(e) {
            var t = typeof e;
            return t == 'string' ||
              t == 'number' ||
              t == 'symbol' ||
              t == 'boolean'
              ? e !== '__proto__'
              : e === null;
          }
          function Hn(e) {
            return !!He && He in e;
          }
          function Rn(e) {
            var t = e && e.constructor,
              n = (typeof t == 'function' && t.prototype) || Q;
            return e === n;
          }
          function Gn(e) {
            return Re.call(e);
          }
          function G(e) {
            if (e != null) {
              try {
                return Be.call(e);
              } catch {}
              try {
                return e + '';
              } catch {}
            }
            return '';
          }
          function Je(e, t) {
            return e === t || (e !== e && t !== t);
          }
          var Dn = ke(
              (function () {
                return arguments;
              })(),
            )
              ? ke
              : function (e) {
                  return K(e) && M.call(e, 'callee') && !qe.call(e, 'callee');
                },
            re = Array.isArray;
          function qn(e) {
            return e != null && Xe(e.length) && !We(e);
          }
          var ge = Wt || Fn;
          function zn(e, t) {
            return Fe(e, t);
          }
          function We(e) {
            if (!Ye(e)) return !1;
            var t = U(e);
            return t == Ce || t == _t || t == A || t == bt;
          }
          function Xe(e) {
            return typeof e == 'number' && e > -1 && e % 1 == 0 && e <= d;
          }
          function Ye(e) {
            var t = typeof e;
            return e != null && (t == 'object' || t == 'function');
          }
          function K(e) {
            return e != null && typeof e == 'object';
          }
          var Ze = $e ? Gt($e) : In;
          function Nn(e) {
            return qn(e) ? Tn(e) : Pn(e);
          }
          function kn() {
            return [];
          }
          function Fn() {
            return !1;
          }
          s.exports = zn;
        })(tt, N)),
      N
    );
  }
  var rt = nt();
  const st = et(rt),
    _e = (s, r, o) => {
      const a = {
        id: [r, o, 'circle'].join('_'),
        source: r,
        type: 'circle',
        paint: { 'circle-color': s, 'circle-radius': 2 },
        filter: ['==', '$type', 'Point'],
      };
      return o && (a['source-layer'] = o), a;
    },
    ye = (s, r, o, a) => {
      const u = {
        id: [o, a, 'polygon'].join('_'),
        source: o,
        type: 'fill',
        paint: {
          'fill-antialias': !0,
          'fill-color': s,
          'fill-outline-color': r,
        },
        filter: ['==', '$type', 'Polygon'],
      };
      return a && (u['source-layer'] = a), u;
    },
    be = (s, r, o) => {
      const a = {
        id: [r, o, 'line'].join('_'),
        source: r,
        layout: { 'line-join': 'round', 'line-cap': 'round' },
        type: 'line',
        paint: { 'line-color': s },
        filter: ['==', '$type', 'LineString'],
      };
      return o && (a['source-layer'] = o), a;
    },
    it = (s, r) => {
      const o = [],
        a = [],
        u = [],
        C = (d) => {
          const g = r.bind(null, d);
          return {
            circle: g(0.8),
            line: g(0.6),
            polygon: g(0.3),
            polygonOutline: g(0.6),
            default: g(1),
          };
        };
      return (
        Object.keys(s).forEach((d) => {
          const g = s[`${d}`];
          if (!g || g.length === 0) {
            const y = C(d);
            a.push(_e(y.circle, d)),
              u.push(be(y.line, d)),
              o.push(ye(y.polygon, y.polygonOutline, d));
          } else
            g.forEach((y) => {
              const A = C(y);
              a.push(_e(A.circle, d, y)),
                u.push(be(A.line, d, y)),
                o.push(ye(A.polygon, A.polygonOutline, d, y));
            });
        }),
        [...o, ...u, ...a]
      );
    },
    ot = (s, r, o) => {
      o = Object.assign({ backgroundColor: '#fff' }, o);
      const a = {
          id: 'background',
          type: 'background',
          paint: { 'background-color': o.backgroundColor },
        },
        u = {};
      return (
        Object.keys(s.sources).forEach((C) => {
          const d = s.sources[`${C}`];
          (d.type === 'vector' || d.type === 'geojson') && (u[`${C}`] = d);
        }),
        { ...s, layers: [a, ...r], sources: u }
      );
    };
  class at {
    _btn;
    elem;
    constructor(r) {
      (r = Object.assign({ show: !0, onToggle: () => {} }, r)),
        (this._btn = this.button()),
        (this._btn.onclick = r.onToggle),
        (this.elem = this.container(this._btn, r.show));
    }
    container(r, o) {
      const a = document.createElement('div');
      return (
        (a.className = 'maplibregl-ctrl maplibregl-ctrl-group'),
        a.appendChild(r),
        o || (a.style.display = 'none'),
        a
      );
    }
    button() {
      const r = document.createElement('button');
      return (
        (r.className = 'maplibregl-ctrl-icon maplibregl-ctrl-inspect'),
        (r.type = 'button'),
        (r.ariaLabel = 'Inspect'),
        r
      );
    }
    setInspectIcon() {
      this._btn.className = 'maplibregl-ctrl-icon maplibregl-ctrl-inspect';
    }
    setMapIcon() {
      this._btn.className = 'maplibregl-ctrl-icon maplibregl-ctrl-map';
    }
  }
  const lt = (s) =>
      typeof s > 'u' || s === null
        ? s
        : s instanceof Date
        ? s.toLocaleString()
        : typeof s == 'object' || typeof s == 'number' || typeof s == 'string'
        ? s.toString()
        : s,
    me = (s, r) =>
      `<div class="maplibregl-inspect-property"><div class="maplibregl-inspect-property-name">${s}</div><div class="maplibregl-inspect-property-value">${lt(
        r,
      )}</div></div>`,
    pt = (s) => `<div class="maplibregl-inspect-layer">${s}</div>`,
    ct = (s) => {
      const r = pt(s.layer['source-layer'] || s.layer.source),
        o = me('$type', s.geometry.type),
        a = Object.keys(s.properties).map((u) => me(u, s.properties[`${u}`]));
      return [r, o].concat(a).join('');
    },
    ut = (s) =>
      s
        .map((r) => `<div class="maplibregl-inspect-feature">${ct(r)}</div>`)
        .join(''),
    ft = (s) => `<div class="maplibregl-inspect-popup">${ut(s)}</div>`,
    ht = (s, r) => {
      let o = 'bright',
        a;
      return (
        /water|ocean|lake|sea|river/.test(s) && (a = 'blue'),
        /state|country|place/.test(s) && (a = 'pink'),
        /road|highway|transport|streets/.test(s) && (a = 'orange'),
        /contour|building/.test(s) && (a = 'monochrome'),
        /building/.test(s) && (o = 'dark'),
        /contour|landuse/.test(s) && (a = 'yellow'),
        /wood|forest|park|landcover|land/.test(s) && (a = 'green'),
        `rgba(${`${Ve({
          luminosity: o,
          hue: a,
          seed: s,
          format: 'rgbArray',
        })},${r || 1}`})`
      );
    },
    gt = (s) =>
      !!(s.metadata && 'maplibregl-inspect:inspect' in Object.keys(s.metadata)),
    dt = (s) => ({
      ...s,
      metadata: Object.assign({}, s.metadata, {
        'maplibregl-inspect:inspect': !0,
      }),
    });
  class W {
    _map;
    _popup;
    _popupBlocked = !1;
    _showInspectMap;
    _originalStyle;
    _toggle;
    options;
    sources;
    assignLayerColor;
    constructor(r) {
      if (!(this instanceof W))
        throw new Error(
          'MaplibreInspect needs to be called with the new keyword',
        );
      let o = null;
      $
        ? (o = new $.Popup({ closeButton: !1, closeOnClick: !1 }))
        : r.popup ||
          console.error(
            'Maplibre GL JS can not be found. Make sure to include it or pass an initialized MaplibreGL Popup to MaplibreInspect if you are using moduleis.',
          ),
        (this.options = {
          showInspectMap: !1,
          showInspectButton: !0,
          showInspectMapPopup: !0,
          showMapPopup: !1,
          showMapPopupOnHover: !0,
          showInspectMapPopupOnHover: !0,
          blockHoverPopupOnClick: !1,
          backgroundColor: '#fff',
          assignLayerColor: ht,
          buildInspectStyle: ot,
          renderPopup: ft,
          popup: o,
          selectThreshold: 5,
          useInspectStyle: !0,
          queryParameters: {},
          sources: {},
          toggleCallback(a) {
            console.log('Inspector status?: ', a);
          },
          ...r,
        }),
        (this.sources = this.options.sources),
        (this.assignLayerColor = this.options.assignLayerColor),
        (this.toggleInspector = this.toggleInspector.bind(this)),
        (this._popup = this.options.popup),
        (this._popupBlocked = !1),
        (this._showInspectMap = this.options.showInspectMap),
        (this._onSourceChange = this._onSourceChange.bind(this)),
        (this._onMouseMove = this._onMouseMove.bind(this)),
        (this._onRightClick = this._onRightClick.bind(this)),
        (this._onStyleChange = this._onStyleChange.bind(this)),
        (this._originalStyle = null),
        (this._toggle = new at({
          show: this.options.showInspectButton,
          onToggle: this.toggleInspector.bind(this),
        }));
    }
    _inspectStyle() {
      let r = this._map?.getStyle();
      if (this._map) {
        const o = it(this.sources, this.assignLayerColor);
        r = this.options.buildInspectStyle(this._map.getStyle(), o, {
          backgroundColor: this.options.backgroundColor,
        });
      }
      return r;
    }
    _onSourceChange(r) {
      const o = this.sources;
      if (this._map) {
        const a = this._map,
          u = a.getStyle(),
          C = Object.keys(u.sources),
          d = Object.assign({}, o);
        r.isSourceLoaded &&
          (Object.keys(a.style.sourceCaches).forEach((g) => {
            const y = a.style.sourceCaches[`${g}`] || { _source: {} },
              A = y._source.vectorLayerIds;
            A
              ? (o[`${g}`] = A)
              : y._source.type === 'geojson' && (o[`${g}`] = []);
          }),
          Object.keys(o).forEach((g) => {
            C.indexOf(g) === -1 && delete o[`${g}`];
          }),
          !st(d, o) && Object.keys(o).length > 0 && this.render());
      }
    }
    _onStyleChange() {
      const r = this._map?.getStyle();
      r && !gt(r) && (this._originalStyle = r);
    }
    _onRightClick() {
      !this.options.showMapPopupOnHover &&
        !this.options.showInspectMapPopupOnHover &&
        !this.options.blockHoverPopupOnClick &&
        this._popup &&
        this._popup.remove();
    }
    _onMouseMove(r) {
      if (this._showInspectMap) {
        if (
          !this.options.showInspectMapPopup ||
          (r.type === 'mousemove' && !this.options.showInspectMapPopupOnHover)
        )
          return;
        r.type === 'click' &&
          this.options.showInspectMapPopupOnHover &&
          this.options.blockHoverPopupOnClick &&
          (this._popupBlocked = !this._popupBlocked);
      } else {
        if (
          !this.options.showMapPopup ||
          (r.type === 'mousemove' && !this.options.showMapPopupOnHover)
        )
          return;
        r.type === 'click' &&
          this.options.showMapPopupOnHover &&
          this.options.blockHoverPopupOnClick &&
          (this._popupBlocked = !this._popupBlocked);
      }
      if (!this._popupBlocked && this._map) {
        let o;
        this.options.selectThreshold === 0
          ? (o = r.point)
          : (o = [
              [
                r.point.x - this.options.selectThreshold,
                r.point.y + this.options.selectThreshold,
              ],
              [
                r.point.x + this.options.selectThreshold,
                r.point.y - this.options.selectThreshold,
              ],
            ]);
        const a =
          this._map.queryRenderedFeatures(o, this.options.queryParameters) ||
          [];
        if (
          ((this._map.getCanvas().style.cursor = a.length ? 'pointer' : ''),
          a.length > 0 && this._popup instanceof $.Popup)
        ) {
          this._popup.setLngLat(r.lngLat);
          const u = this.options.renderPopup(a);
          typeof u == 'string'
            ? this._popup.setHTML(u)
            : this._popup.setDOMContent(u),
            this._popup.addTo(this._map);
        } else this._popup?.remove();
      }
    }
    toggleInspector() {
      (this._showInspectMap = !this._showInspectMap),
        this.options.toggleCallback(this._showInspectMap),
        this.render();
    }
    render() {
      if (this._showInspectMap) {
        if (this.options.useInspectStyle) {
          const r = this._inspectStyle();
          this._map?.setStyle(dt(r));
        }
        this._toggle.setMapIcon();
      }
      !this._showInspectMap &&
        this._originalStyle &&
        (this.options.useInspectStyle &&
          this._map?.setStyle(this._originalStyle),
        this._popup && this._popup.remove(),
        this._toggle.setInspectIcon());
    }
    onAdd(r) {
      return (
        (this._map = r),
        Object.keys(this.sources).length === 0 &&
          r.on('sourcedata', this._onSourceChange),
        r.on('styledata', this._onStyleChange),
        r.on('load', this._onStyleChange),
        r.on('mousemove', this._onMouseMove),
        r.on('click', this._onMouseMove),
        r.on('contextmenu', this._onRightClick),
        this._toggle.elem
      );
    }
    onRemove() {
      this._map?.off('styledata', this._onStyleChange),
        this._map?.off('load', this._onStyleChange),
        this._map?.off('sourcedata', this._onSourceChange),
        this._map?.off('mousemove', this._onMouseMove),
        this._map?.off('click', this._onMouseMove),
        this._map?.off('contextmenu', this._onRightClick);
      const r = this._toggle.elem;
      r.parentNode?.removeChild(r), (this._map = void 0);
    }
  }
  (O.MaplibreInspect = W),
    (O.default = W),
    Object.defineProperties(O, {
      __esModule: { value: !0 },
      [Symbol.toStringTag]: { value: 'Module' },
    });
});
